downloader/src/main/kotlin/vcs/GitRepo.kt

```diff
@@ -58,6 +59,28 @@ class GitRepo : GitBase() {
                     return super.getInfo().copy(path = manifestFile.relativeTo(workingDir).invariantSeparatorsPath)
                 }
 
+                override fun getNested(): Map<String, VcsInfo> {
+                    val paths = runRepoCommand(workingDir, "list", "-p").stdout.lines().filter {
+                        // Skip lines which belong to the upgrade message.
+                        it.isNotBlank() && !it.startsWith("...") && !it.startsWith(" ")
+                    }
+
+                    val nested = mutableMapOf<String, VcsInfo>()
+
+                    paths.forEach { path ->
+                        // Add the nested Repo project.
+                        val workingTree = VersionControlSystem.forDirectory(File(getRootPath(), path))
+                        nested[path] = workingTree?.getInfo() ?: VcsInfo.EMPTY
+
+                        // Add the Git submodules of the nested Repo project.
+                        workingTree?.getNested()?.forEach { (submodulePath, vcsInfo) ->
+                            nested["$path/$submodulePath"] = vcsInfo
```

Comment:
Yes, I think we should have the full picture. I don't see a reason to exclude recursive submodules.
Not sure if we have to handle loops, does Git allow that? If there is a loop, shouldn't this fail while cloning because you cannot have infinite directories?

Reply:
Right, with pure Git / submodules we can leave the loop-issue to Git. I was thinking about the case when git-repo "closes the loop". But that cannot happen, because even if a nested submodule would be the manifest repo, it would be closed with pure Git, not with git-repo. So we're good.